### PR TITLE
feat(website/code component): add copy to clipboard to all code sample

### PR DESCRIPTION
**Pull request by mohamedMok**
## I have read [the contributing guidelines](https://mozaic.adeo.cloud/Contributing/)

- [X] Yes
- [ ] No

## Does this PR introduce a breaking change?

read [What is a breaking change ?](https://mozaic.adeo.cloud/Contributing/Prerequisite/GitConventions/#breaking-changes-)

- [ ] Yes
- [X] No

## Describe the changes

What: Add copy to clipboard to all code sample (even inline ones)
How: Link customised component (Code, InlineCode) to mdx element code. 
Refactor code to have only one function copy to clipboard for all browsers

Github issue number or Jira issue URL: https://design-system-adeo.atlassian.net/browse/MZC-48

### CODE REVIEW

File: packages/gatsby-theme-styleguide/src/gatsby-components/Code/Code.js [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    params.split(`&`).reduce((merged, param) => {
+      const [key, value] = param.split(`=`)
+      merged[key] = value
+      return merged
+    }, {})
+  )
+}
+
+const copyToClipboardClick = (str, setCopied) => {
+  setCopied(true)
```
Comment (gael-boyenval) [inline packages/gatsby-theme-styleguide/src/gatsby-components/Code/Code.js]: toggleCopy() ?

File: packages/gatsby-theme-styleguide/src/gatsby-components/Code/Code.js [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+`
+
+const Body = styled.div`
+  flex: 1;
+  overflow: auto;
+  position: relative;
+  border-left: solid 1px #ececec;
+`
+
+const getParams = (name = ``) => {
```
Comment (gael-boyenval) [inline packages/gatsby-theme-styleguide/src/gatsby-components/Code/Code.js]: getLangFromClassName

File: packages/gatsby-theme-styleguide/src/gatsby-components/Code/Code.js [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+  const [lang, params = ``] = name.split(`:`)
```
Comment (gael-boyenval) [inline packages/gatsby-theme-styleguide/src/gatsby-components/Code/Code.js]: comment to show the class structure

File: packages/gatsby-theme-styleguide/src/gatsby-components/Code/Code.js [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+ */
+const Code = ({
+  children,
+  className = children.props ? children.props.className : ``,
+  copy = true,
+  isOpen,
+  fullScreen,
+}) => {
+  const [copied, setCopied] = useState(false)
+  const [language] = getParams(className)
```
Comment (gael-boyenval) [inline packages/gatsby-theme-styleguide/src/gatsby-components/Code/Code.js]: restructuring not necessary.

File: packages/gatsby-theme-styleguide/src/gatsby-components/Code/normalize.js [hunk 5]
```diff
@@ -0,0 +1,155 @@
+const COMMENT_START = new RegExp(`(#|\\/\\/|\\{\\/\\*|\\/\\*+|<!--)`)
```
Comment (gael-boyenval) [inline packages/gatsby-theme-styleguide/src/gatsby-components/Code/normalize.js]: use a link to tell where does it come from

File: packages/gatsby-theme-styleguide/src/utils/copy-to-clipboard.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
+const copyToClipboard = str => {
```
Comment (gael-boyenval) [inline packages/gatsby-theme-styleguide/src/utils/copy-to-clipboard.js]: Test here if the browser is compatible with the copy API your components to manage state. 
You can catch errors and reject to let the component make their own error handling. 

inside your component the usage should look something like : 

```js
copyToClipBoard(somethingToCopy)
  .then(() => setSate(...))
  .catch(e => handleError(...))
```

we could ba able to remove a lot of duplication in the component using it

File: packages/gatsby-theme-styleguide/src/gatsby-components/Code/normalize.js (hunk 5 above)
Comment (mohamedMok) [reply packages/gatsby-theme-styleguide/src/gatsby-components/Code/normalize.js]: Added some comment to explain, what and where it comes from

File: packages/gatsby-theme-styleguide/src/utils/copy-to-clipboard.js (hunk 6 above)
Comment (mohamedMok) [reply packages/gatsby-theme-styleguide/src/utils/copy-to-clipboard.js]: Added the catch to the function

File: packages/gatsby-theme-styleguide/src/gatsby-components/Code/Code.js (hunk 4 above)
Comment (mohamedMok) [reply packages/gatsby-theme-styleguide/src/gatsby-components/Code/Code.js]: Removed

File: packages/gatsby-theme-styleguide/src/gatsby-components/Code/Code.js [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+const Body = styled.div`
+  flex: 1;
+  overflow: auto;
+  position: relative;
+  border-left: solid 1px #ececec;
+`
+
+const copyToClipboardClick = (str, toogleCopy) => {
+  toogleCopy(true)
```
Comment (gael-boyenval) [inline packages/gatsby-theme-styleguide/src/gatsby-components/Code/Code.js]: Try again one last time ;)

You want to toggle the state only if the copy worked, if not, return null or false. 
No console logs in front-end code. 

if there is an error, your state will stay `copied: true` which is not what we want.

**Comment by gael-boyenval** [comment]
Capture d’écran 2019-10-29 à 15 54 31

File: packages/gatsby-theme-styleguide/src/gatsby-components/Code/Code.js (hunk 7 above)
Comment (mohamedMok) [reply packages/gatsby-theme-styleguide/src/gatsby-components/Code/Code.js]: 👍  Got it! Almost there :)

**Pull request merged**